### PR TITLE
Pass on the GPU acceleration section of Chapter 13

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR just cuts or moves to footnote any text it can. I also rearrange the code a bit to avoid redundancy about not needing to do a Skia -> SDL copy.

### CODE REVIEW

**Comment by pavpanchekha** [comment]
This branch is marked to merge into `ch13-pass` which is PR #505. That's just for ease of reviewing. If this is approved I'll just keep it around until everything is merged.

File: book/animations.md [hunk 1]
```diff
@@ -192,32 +192,81 @@ decade of the [2010s][cssanim-hist].
...
+
+SDL and Skia support all of these steps---in fact, it's mostly a
+matter of passing them the right parameters. Let's do that. Note that
+a real browser typically implements both CPU and GPU raster and draw,
+because in some cases CPU raster and draw can be faster than using the
+GPU.[^example-cpu-fast] In our browswer, for simplicity, we'll stick
+to GPU mode for all pages.
+
+[^example-cpu-fast]: For example, the upload step isn't necessary on a
+    CPU, nor is the upload step. And GPUs typically have a lot less
```
Comment (chrishtr) [inline book/animations.md:245]:
```suggestion
    CPU, nor is the compile step. And GPUs typically have a lot less
```

File: book/animations.md [hunk 2]
```diff
@@ -378,20 +349,40 @@ class Browser:
         sdl2.SDL_GL_SwapWindow(self.sdl_window)
 ```
 
-Let's go back and test the [counter example](scheduling.md#animating-frames)
-from Chapter 12. Without GPU, the results were:
+Finally, in our browser also creates a Skia surfaces for the
```
Comment (chrishtr) [inline book/animations.md:352]:
```suggestion
Finally, our browser also creates a Skia surfaces for the
```

File: book/animations.md [hunk 3]
```diff
@@ -239,81 +288,28 @@ class Browser:
...
-leaf textures are generated from raw display list data structures, and internal
-surface tree textures are generated by computing visual effects on one or more
-children. The root of the tree is the framebuffer texture.
-
-[^gpu-texture]: Recall from [Chapter 11](visual-effects.md) that there can be
-surfaces that draw into other surfaces, forming a tree. Skia internally does
-this sometimes, based on various triggers such as blend modes. The internal
-nodes draw into intermediate textures.
-
-The time to run GPU raster is roughly the sum of the time for these four steps.
```
Comment (chrishtr) [inline book/animations.md:299]: I think this paragraph is good to keep, as it explains how slowdowns can come from each part. I suppose we don't actually use this knowledge though. Footnote?

File: book/animations.md [hunk 4]
```diff
@@ -192,32 +192,81 @@ decade of the [2010s][cssanim-hist].
...
+
+With 66ms per frame, our browser is barely doing fifteen frames per
+second, not sixty. For smooth animations we'll need to speed up raster
+and draw.
+
+The best way to do that is to move raster and draw to the [GPU][gpu].
+A GPU is essentially a chip in your computer that runs programs, much
+like your CPU. But the GPU is specialized toward running very simple
+programs with massive parallelism; it was developed to do simple
+operations to every pixel on the screen. This makes GPUs faster for
```
Comment (chrishtr) [inline book/animations.md:211]:
```suggestion
operations, in parallel, for every pixel on the screen. This makes GPUs faster for
```

File: book/animations.md [hunk 5]
```diff
@@ -192,32 +192,81 @@ decade of the [2010s][cssanim-hist].
...
+    Time in render on average: 23ms
+
+With 66ms per frame, our browser is barely doing fifteen frames per
+second, not sixty. For smooth animations we'll need to speed up raster
+and draw.
+
+The best way to do that is to move raster and draw to the [GPU][gpu].
+A GPU is essentially a chip in your computer that runs programs, much
+like your CPU. But the GPU is specialized toward running very simple
+programs with massive parallelism; it was developed to do simple
```
Comment (chrishtr) [inline book/animations.md:210]:
```suggestion
programs with massive parallelism; it was developed to apply simple
```

File: book/animations.md (hunk 3 above)
Comment (pavpanchekha) [reply book/animations.md:297]: I merged this text into the `example-cpu-fast` footnote.

File: book/animations.md [hunk 6]
```diff
@@ -190,32 +190,85 @@ decade of the [2010s][cssanim-hist].
...
+[^compiled-gpu]: That's right, GPU programs are dynamically compiled! This
+allows GPU programs to be portable across a wide variety of implementations
+that may have very different instruction sets or acceleration tactics.
+Of course, between frames these compiled programs will typically be
+cached, so this step won't occur on every frame.
+
+* *Raster* every drawing command into GPU textures.[^texture]
+
+[^texture]: A surface represented on the GPU is called a *texture*. There can be
+more than one surface, and practically speaking they often can't be rastered in
```
Comment (chrishtr) [inline book/animations.md:230]:
```suggestion
more than one texture, and practically speaking they often can't be rastered in
```

File: book/animations.md [hunk 7]
```diff
@@ -190,32 +190,85 @@ decade of the [2010s][cssanim-hist].
...
+parallel with each other.
+
+* *Draw* the textures onto the screen.
+
+SDL and Skia support all of these steps---in fact, it's mostly a
+matter of passing them the right parameters. Let's do that. Note that
+a real browser typically implements both CPU and GPU raster and draw,
+because in some cases CPU raster and draw can be faster than using the
+GPU.[^example-cpu-fast] In our browswer, for simplicity, we'll stick
```
Comment (chrishtr) [inline book/animations.md:239]:
```suggestion
GPU.[^example-cpu-fast] In our browser, for simplicity, we'll stick
```

File: book/animations.md [hunk 8]
```diff
@@ -190,32 +190,85 @@ decade of the [2010s][cssanim-hist].
...
+to GPU mode for all pages.
+
+[^example-cpu-fast]: Any of the four steps can make GPU raster and
+ draw slow. Large display list take a while to upload. Complex
```
Comment (chrishtr) [inline book/animations.md:243]:
```suggestion
 draw slow. Large display lists take a while to upload. Complex
```

File: book/animations.md [hunk 9]
```diff
@@ -376,20 +351,40 @@ class Browser:
...
+should be all that's necessary for our browser to raster and draw on
+the GPU. And as expected, speed is much improved:
 
     Time in raster-and-draw on average: 24ms
     Time in render on average: 23ms
 
-So GPU acceleration speeds up raster-and-draw by more than 60%. (If you're on a
-computer with a non-virtualized GL driver you will probably see even more
-speedup than that.)
+That's almost three times faster, almost fast enough to hit sixty
```
Comment (chrishtr) [inline book/animations.md:383]:
```suggestion
That's almost three times faster, almost fast enough to hit sixty
```
```suggestion
That's almost three times faster, almost fast enough to hit sixty
```
```suggestion
That's about three times faster, almost fast enough to hit sixty
```

File: book/animations.md [hunk 10]
```diff
@@ -376,20 +351,40 @@ class Browser:
...
+frames per second. And on your computer, you'll likely see even more
+speedup than that, so perhaps your animations are already quite
+smooth. But if we want to go faster yet, we'll need to reduce the
+total amount of raster and draw work we're doing.
```
Comment (chrishtr) [inline book/animations.md:387]:
```suggestion
total amount of work in raster and draw.
```

**Pull request merged**